HTTP Origin header — improve overview (#10709)
mdn/content: files/en-us/web/http/headers/origin/index.md

---
{{HTTPSidebar}}

> The **`Origin`** HyperText Transfer Protocol (HTTP) request header indicates the {{glossary("origin")}} (scheme, hostname, and port) that _caused_ the request.

Reviewer: ```suggestion
The **`Origin`** HTTP request header indicates the {{glossary("origin")}} (scheme, hostname, and port) that _caused_ the request.
```
(I know you didn’t add that text in this PR — somebody else did some weeks back — but it’s not necessary, so let’s just drop it while we’re in here.

```
Origin: null
Origin: <scheme>://<hostname>
Origin: <scheme>://<hostname>:<port>
```

## Directives

- `null`

>   - : The origin is "privacy sensitive", or is an _opaque origin_ as defined by the HTML specification (specific cases are listed in the [description](#description) section).

Author: I am a little concerned about this. In https://www.rfc-editor.org/rfc/rfc6454#section-7.3 this is mostly about "privacy sensitivity". However the HTML spec talks more about null as an internal value. I have chosen not to mention the fetch spec global id at all.

I have decided to state the rules and then forward link down to the description for more cases. In summary though, the way I see it though is that it is `null` if you _can't_ send the origin (i.e. it doesn't have something that quite matches the triple as defined), or if you don't want to send the origin because it isn't needed and we try reduce the information that might potentially have privacy implications unless needed.

Reviewer: I think the less that’s said about trying to explain what a null origin signifies, the better. I think attempting it would mostly just add distraction and raise more questions/perplexity from readers.

- `<scheme>`
  - : The protocol that is used.
    Usually, it is the HTTP protocol or its secured version, HTTPS.
- `<hostname>`
  - : The domain name or the IP address of the origin server.
- `<port>` {{optional_inline}}
  - : Port number on which the server is listening.
    If no port is given, the default port for the requested service is implied (e.g., "80" for an HTTP URL) .

## Description

> The `Origin` header is similar to the {{HTTPHeader("Referer")}} header, but does not disclose the whole path, and may be `null`.

Reviewer: ```suggestion
The `Origin` header is similar to the {{HTTPHeader("Referer")}} header, but does not disclose the path, and may be `null`.
```
It doesn’t expose any path part at all

> The `Origin` is similar to the {{HTTPHeader("Referer")}} header, but does not disclose the whole path, and may contain `null`.
> It is used to provide the "security context" for the origin request, except in cases where the origin information would be sensitive or unnecessary.

Reviewer: ```suggestion
The `Origin` header is similar to the {{HTTPHeader("Referer")}} header, but does not disclose the whole path, and may be `null`.
It is used to provide the "security context" for the origin request, except in cases where the origin information would be sensitive or unnecessary.
```

Broadly speaking, user agents add the {{httpheader("Origin")}} request header to:
- {{Glossary("CORS", "cross origin")}} requests.
- [same-origin](/en-US/docs/Web/Security/Same-origin_policy) requests except for {{HTTPMethod("GET")}} or {{HTTPMethod("HEAD")}} requests (i.e. they are added to same-origin {{HTTPMethod("POST")}}, {{HTTPMethod("OPTIONS")}}, {{HTTPMethod("PUT")}}, {{HTTPMethod("PATCH")}}, and {{HTTPMethod("DELETE")}} requests).

There are some exceptions to the above rules; for example, if a cross-origin {{HTTPMethod("GET")}} or {{HTTPMethod("HEAD")}} request is made in [no-cors mode](/en-US/docs/Web/API/Request/mode#value), the `Origin` header will not be added.

> The `Origin` header may return `null` in a number of cases, including (non-exhaustively):

Reviewer: ```suggestion
The `Origin` header value may be `null` in a number of cases, including (non-exhaustively):
```

> - Origin's whose scheme is not one of `http`, `https`, `ftp`, `ws`, `wss`, or `gopher` (including `blob`, `file` and `data`).
> - Cross origin images and media data, including that in `<img>`, `<video>` and `<audio>` elements.
> - Documents created programmatically using `createDocument()`, generated from a `data:` url, or that do not have a creator browsing context. 
> - Redirects across origins.
> - iframes with a sandbox attribute that doesn’t contain the value `allow-same-origin`.
> - Responses that are network errors.
> 
> > **Note:** There is a more detailed listing of case that may return `null` on Stack Overflow here: [When do browsers send the Origin header? When do browsers set the origin to null?](https://stackoverflow.com/a/42242802/).

Reviewer: ```suggestion
- Origins whose scheme is not one of `http`, `https`, `ftp`, `ws`, `wss`, or `gopher` (including `blob`, `file` and `data`).
- Cross-origin images and media data, including that in `<img>`, `<video>` and `<audio>` elements.
- Documents created programmatically using `createDocument()`, generated from a `data:` url, or that do not have a creator browsing context. 
- Redirects across origins.
- iframes with a sandbox attribute that doesn’t contain the value `allow-same-origin`.
- Responses that are network errors.

> **Note:** There is a more detailed listing of case that may return `null` on Stack Overflow here: [When do browsers send the Origin header? When do browsers set the origin to null?](https://stackoverflow.com/a/42242802/).
```
